app/src/org/commcare/tasks/FormRecordToFileTask.java

```diff
@@ -0,0 +1,278 @@
+                    Log.d(TAG, "Copying file: " + file.getName() + " failed with: " + ie.getMessage());
+                    publishProgress(("File writing failed: " + ie.getMessage()));
+                    return FormUploadUtil.FAILURE;
+                }
+            }
+        }
+        // write any form.properties we want
+        writeProperties(myDir);
+        return FormUploadUtil.FULL_SUCCESS;
+    }
+
+    /**
+     * Writes any properties of this form/user the receiving tablet might want to form.properties
+     * Current properties:
+     *  PostURL:    The receiver will attempt to submit to this URL instead of its default URL.
+     *              We do this because HQ uses the receiver URL to help display forms prettily.
+     * @param formInstanceFolder: the form instance folder to write in
+     */
+    private void writeProperties(File formInstanceFolder) {
+        FileOutputStream outputStream = null;
+        try {
+            File formProperties = new File(formInstanceFolder, "form.properties");
+            outputStream = new FileOutputStream(formProperties);
+            Properties properties = new Properties();
+            SharedPreferences settings = CommCareApplication._().getCurrentApp().getAppPreferences();
+            // HQ likes us to submit forms to the "correct" app and user specific URL
+            String postUrl = settings.getString(CommCarePreferences.PREFS_SUBMISSION_URL_KEY,
+                    c.getString(R.string.PostURL));
+            properties.setProperty("PostURL", postUrl);
+            properties.store(outputStream, null);
+        } catch(IOException e){
+            // we'll just ignore this, not the end of the world
+            e.printStackTrace();
+        } finally{
+            if(outputStream != null){
+                try {
+                    outputStream.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+    }
+
+    @Override
+    protected Pair<Long, FormRecord[]> doTaskBackground(String... params) {
+        Log.d(TAG, "Doing zip task in background with params: " + params);
+
+        // we want this directory to be clean
+        if (storedFormDirectory.exists()) {
+            storedFormDirectory.delete();
+        }
+        storedFormDirectory.mkdirs();
+
+        SqlStorage<FormRecord> storage = CommCareApplication._().getUserStorage(FormRecord.class);
+        Vector<Integer> ids = StorageUtils.getUnsentOrUnprocessedFormsForCurrentApp(storage);
+
+        if (ids.size() > 0) {
+            FormRecord[] records = new FormRecord[ids.size()];
+            for (int i = 0; i < ids.size(); ++i) {
+                records[i] = storage.read(ids.elementAt(i).intValue());
+            }
+
+            results = new Long[records.length];
+            for (int i = 0; i < records.length; ++i) {
+                //Assume failure
+                results[i] = FormUploadUtil.FAILURE;
+            }

```

Comment:
This loop isn't needed if you move the work into the loop above. They should iterate over the same range.